pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -247,6 +267,17 @@ public void addSegment(ImmutableSegmentImpl segment, @Nullable ThreadSafeMutable
     Lock segmentLock = SegmentLocks.getSegmentLock(_tableNameWithType, segmentName);
     segmentLock.lock();
     try {
+      // Skip adding segments that has segment EndTime in the comparison cols earlier than (largestSeenTimestamp - TTL).
```

Comment:
(MAJOR) Just realized a problem. If we return here, the valid doc ids and queryable doc ids won't be set in the segment. The correct solution should be:
- If `delete` is not enabled, set valid doc ids snapshot into the segment
- If `delete` is enabled, since we don't have snapshot for queryable doc ids, we need to loop over valid doc ids to create the queryable doc ids (find non-delete records)

Can you add a test to catch this problem? With current solution, after restarting all the invalid docs will show up again